python/lsst/daf/butler/direct_query_driver/_driver.py

```diff
@@ -0,0 +1,1200 @@
+        result_page = cursor.next()
+        if result_page.next_key is not None:
+            self._cursors[result_page.next_key] = cursor
+        return result_page
+
+    def materialize(
+        self,
+        tree: qt.QueryTree,
+        dimensions: DimensionGroup,
+        datasets: frozenset[str],
+    ) -> qt.MaterializationKey:
+        # Docstring inherited.
+        if self._exit_stack is None:
+            raise RuntimeError("QueryDriver context must be entered before 'materialize' is called.")
+        _, builder = self.build_query(tree, qt.ColumnSet(dimensions))
+        # Current implementation ignores 'datasets' aside from remembering
+        # them, because figuring out what to put in the temporary table for
+        # them is tricky, especially if calibration collections are involved.
+        # That's okay because:
+        #
+        # - the query whose results we materialize includes the dataset
+        #   searches as constraints;
+        #
+        # - we still (in Query.materialize) join the dataset searches back in
+        #   anyway, and given materialized data IDs the join to the dataset
+        #   search is straightforward and definitely well-indexed, and not much
+        #   (if at all) worse than joining back in on a materialized UUID.
+        #
+        sql_select = builder.select()
+        table = self._exit_stack.enter_context(self.db.temporary_table(builder.make_table_spec()))
+        self.db.insert(table, select=sql_select)
+        key = uuid.uuid4()
+        self._materializations[key] = (table, datasets, builder.postprocessing)
+        return key
+
+    def upload_data_coordinates(
+        self, dimensions: DimensionGroup, rows: Iterable[tuple[DataIdValue, ...]]
+    ) -> qt.DataCoordinateUploadKey:
+        # Docstring inherited.
+        if self._exit_stack is None:
+            raise RuntimeError(
+                "QueryDriver context must be entered before 'upload_data_coordinates' is called."
+            )
+        columns = qt.ColumnSet(dimensions).drop_implied_dimension_keys()
+        table_spec = ddl.TableSpec(
+            [columns.get_column_spec(logical_table, field).to_sql_spec() for logical_table, field in columns]
+        )
+        dict_rows: list[dict[str, Any]]
+        if not columns:
+            table_spec.fields.add(
+                ddl.FieldSpec(
+                    QueryBuilder.EMPTY_COLUMNS_NAME, dtype=QueryBuilder.EMPTY_COLUMNS_TYPE, nullable=True
+                )
+            )
+            dict_rows = [{QueryBuilder.EMPTY_COLUMNS_NAME: None}]
+        else:
+            dict_rows = [dict(zip(dimensions.required, values)) for values in rows]
+        from_clause: sqlalchemy.FromClause
+        if len(dict_rows) > self._constant_rows_limit:
+            from_clause = self._exit_stack.enter_context(self.db.temporary_table(table_spec))
+            self.db.insert(from_clause, *dict_rows)
```

Comment:
I think mainly it would let the planner know min/max values for the columns, which might affect its decisions on join order etc.  But whether the planner having that knowledge outweighs the cost of the analyze is something we'd have to benchmark probably.